### ISSUE TITLE
get session value in socketio not working

### ISSUE THREAD

**Issue by Pang654** [miguelgrinberg/Flask-SocketIO#1256]
Hi Miguel,

Thank you for all the help for these few days.  I have a problem again.  I am trying to get the session value in the socketio, but it does not work. It works in the HTTP request.  I tried to print the session email which I have set before, but it printed "None"
I have tried changing the manage_session to True. Both True and False gave "None".

Code for server
```
import eventlet
eventlet.monkey_patch()

from flask import Flask , jsonify, session
from flask_socketio import SocketIO, emit, join_room, leave_room, send, ConnectionRefusedError
from flask_session import Session

app = Flask(__name__)

app_plain_text = "Meow Meow"

#Session configurations
app.config['SECRET_KEY'] = "Meow Meow"
app.config['SESSION_TYPE'] = "redis"
Session(app)

socketio = SocketIO(app, manage_session=False, cors_allowed_origins="*", logger=True, engineio_logger=True)


app.config['DEBUG'] = True

def ack():
  print('message was received!')

# testing use only

@app.route("/test", methods=["POST"])
def test():
  session["email"] = 'test@test.com'
  return jsonify({"result": "successful"})

@app.route("/check_session", methods=["POST"])
def check_session():
  email = session.get("email")
  print(email)
  return jsonify({"result": "successful"})


@socketio.on('connect')
def test_connect():
  email = session.get("email")
  print(email)
  print('Client connected')

@socketio.on('disconnect')
def test_disconnect():
  print('Client disconnected')

if __name__ == "__main__":
  socketio.run(app, host='0.0.0.0', port=5000)
```

Server Log
```
Server initialized for eventlet.
 * Restarting with stat
Server initialized for eventlet.
 * Debugger is active!
 * Debugger PIN: 281-591-133
(24663) wsgi starting up on http://0.0.0.0:5000
(24663) accepted (', 49548)
 - - [19/Apr/2020 23:18:08] "POST /test HTTP/1.1" 200 295 0.012531
(24663) accepted ('', 49560)
test@test.com
 - - [19/Apr/2020 23:18:45] "POST /check_session HTTP/1.1" 200 295 0.008553
(24663) accepted (, 49562)
b80acb19244246fe9a69b7cfecc10324: Sending packet OPEN data {'sid': 'b80acb19244246fe9a69b7cfecc10324', 'upgrades': ['websocket'], 'pingTimeout': 60000, 'pingInterval': 25000}
None
Client connected
b80acb19244246fe9a69b7cfecc10324: Sending packet MESSAGE data 0
 - - [19/Apr/2020 23:18:58] "GET /socket.io/?transport=polling&EIO=3&t=1587309538.6125052 HTTP/1.1" 200 349 0.008657
(24663) accepted (', 49563)
b80acb19244246fe9a69b7cfecc10324: Received request to upgrade to websocket
b80acb19244246fe9a69b7cfecc10324: Upgrade to websocket successful
b80acb19244246fe9a69b7cfecc10324: Received packet PING data None
b80acb19244246fe9a69b7cfecc10324: Sending packet PONG data None
Client disconnected
 - - [19/Apr/2020 23:19:16] "GET /socket.io/?transport=websocket&EIO=3&sid=b80acb19244246fe9a69b7cfecc10324&t=1587309538.7035189 HTTP/1.1" 200 0 
```

Client Log
```
Polling connection accepted with {'sid': 'b80acb19244246fe9a69b7cfecc10324', 'upgrades': ['websocket'], 'pingTimeout': 60000, 'pingInterval': 25000}
Engine.IO connection established
Received packet MESSAGE data 0
Namespace / is connected
Attempting WebSocket upgrade to ws://mynetgear.com:5000/socket.io/?transport=websocket&EIO=3
I'm connected!
WebSocket upgrade was successful
Sending packet PING data None
Received packet NOOP data None
Received packet PONG data None
Sending packet CLOSE data None
Engine.IO connection dropped
Exiting write loop task
I'm disconnected!
Received packet NOOP data None
Waiting for write loop task to end
Waiting for ping loop task to end
Exiting ping task
```
pip freeze
```
astroid==2.3.3
certifi==2020.4.5.1
chardet==3.0.4
click==7.1.1
dnspython==1.16.0
eventlet==0.25.2
Flask==1.1.2
Flask-Session==0.3.1
Flask-SocketIO==4.2.1
gevent==1.5.0
greenlet==0.4.15
idna==2.9
isort==4.3.21
itsdangerous==1.1.0
Jinja2==2.11.2
lazy-object-proxy==1.4.3
MarkupSafe==1.1.1
mccabe==0.6.1
monotonic==1.5
pkg-resources==0.0.0
pylint==2.4.4
python-engineio==3.12.1
python-socketio==4.5.1
redis==3.4.1
requests==2.23.0
six==1.14.0
urllib3==1.25.8
websocket-client==0.57.0
Werkzeug==1.0.1
wrapt==1.11.2
```

**Comment by miguelgrinberg** [comment]
Can I see your client?

**Comment by Pang654** [comment]
For the http request, im using postman to do the testing. I am using simple python socketio for the socketio to do the testing. 
```
import socketio

# standard Python
sio = socketio.Client(logger=True, engineio_logger=True)

@sio.event
def connect():
  print("I'm connected!")

@sio.event
def disconnect():
  print("I'm disconnected!")


if __name__ == "__main__":
  sio.connect('http://myserver.com:5000')
```

**Comment by miguelgrinberg** [comment]
Wait a second. How did you expect the session was going to be shared if you are using two different clients for the HTTP and SocketIO parts? User sessions are per client, that's the whole point of having them.

**Comment by Pang654** [comment]
I have build a simple http request, so the HTTP clients and socketio clients are on the same page. The result is still the same. 

Client Side code

```
import time

import socketio
import requests

# standard Python
sio = socketio.Client(logger=True, engineio_logger=True)

@sio.event
def connect():
  print("I'm connected!")

@sio.event
def disconnect():
  print("I'm disconnected!")


if __name__ == "__main__":
  s = requests.Session()

  test_url = "http://carlhung.mynetgear.com/test"
  headers = {'Accept': '*/*',
             'X-Requested-With': 'XMLHttpRequest',
             'Content-Type': 'application/json',
             }
  test = s.request(method='POST', url=test_url, headers=headers)
  print(test.text)

  check_session_url = "http://mynetgear.com/check_session"
  headers = {'Accept': '*/*',
             'X-Requested-With': 'XMLHttpRequest',
             'Content-Type': 'application/json',
             }
  check_session = s.request(method='POST', url=check_session_url, headers=headers)
  print(check_session.text)

  time.sleep(10)

  sio.connect('http://mynetgear.com:5000')

```

Client Log
```
{
  "result": "successful"
}

{
  "result": "successful"
}

Attempting polling connection to http://carlhung.mynetgear.com:5000/socket.io/?transport=polling&EIO=3
Polling connection accepted with {'sid': '320e7eee32bf404a890740de1b273b5f', 'upgrades': ['websocket'], 'pingTimeout': 60000, 'pingInterval': 25000}
Engine.IO connection established
Received packet MESSAGE data 0
Namespace / is connected
Attempting WebSocket upgrade to ws://mynetgear.com:5000/socket.io/?transport=websocket&EIO=3
I'm connected!
WebSocket upgrade was successful
Sending packet PING data None
Received packet NOOP data None
Received packet PONG data None
Sending packet CLOSE data None
Engine.IO connection dropped
Exiting write loop task
```

Server Log
```
Server initialized for eventlet.
 * Restarting with stat
Server initialized for eventlet.
 * Debugger is active!
 * Debugger PIN: 281-591-133
(3426) wsgi starting up on http://0.0.0.0:5000
(3426) accepted ('61.93.48.188', 51285)
[20/Apr/2020 12:52:53] "POST /test HTTP/1.1" 200 276 0.010528
test@test.com
 [20/Apr/2020 12:52:53] "POST /check_session HTTP/1.1" 200 276 0.006611
(3426) accepted ('61.93.48.188', 51287)
320e7eee32bf404a890740de1b273b5f: Sending packet OPEN data {'sid': '320e7eee32bf404a890740de1b273b5f', 'upgrades': ['websocket'], 'pingTimeout': 60000, 'pingInterval': 25000}
None
Client connected
320e7eee32bf404a890740de1b273b5f: Sending packet MESSAGE data 0
[20/Apr/2020 12:53:03] "GET /socket.io/?transport=polling&EIO=3&t=1587358383.273586 HTTP/1.1" 200 349 0.008756
(3426) accepted ('61.93.48.188', 51288)
320e7eee32bf404a890740de1b273b5f: Received request to upgrade to websocket
320e7eee32bf404a890740de1b273b5f: Upgrade to websocket successful
320e7eee32bf404a890740de1b273b5f: Received packet PING data None
320e7eee32bf404a890740de1b273b5f: Sending packet PONG data None
320e7eee32bf404a890740de1b273b5f: Received packet CLOSE data None
Client disconnected
[20/Apr/2020 12:53:10] "GET /socket.io/?transport=websocket&EIO=3&sid=320e7eee32bf404a890740de1b273b5f&t=1587358383.319614 HTTP/1.1" 200 0 6.754144
```

**Comment by miguelgrinberg** [comment]
There are two problems with your example:

- You are still using two different clients for your HTTP and Socket.IO traffic. Write a JavaScript test for the browser, that's the easiest way to share the same client for both.
- Your HTTP and Socket.IO servers are on different domains or ports. Cookies are not shared across different domains or ports.

**Comment by Pang654** [comment]
Hi,

I have created the javascript test for the browser, it works fine. But when I use flutter mobile app, it doesn't work. I am quite new to socketio. May I ask is it flutter socketio client side doesnt allow sharing session between HTTP and WebSocket?

**Comment by miguelgrinberg** [comment]
I have never used your socket.io client. The key to make this work is that the cookie set by the HTTP transaction is used when you connect via Socket.IO. If your HTTP and Socket.IO libraries do not share a session, maybe you can get the cookie from the HTTP library and set it in the connect call for Socket.IO.

**Comment by Pang654** [comment]
yup trying to do the same thing. Thanks for the help.

**Comment by Pang654** [comment]
Hi,
I found out that the flutter socktio library is pretty new. there is no option for us to manual send the sid in the connect call. I can only emit, and send it as a data. is there any methods that I can update the sid after I emit the sid to the server.

**Comment by miguelgrinberg** [comment]
You shouldn't need to send a sid manually. The server assigned sids to clients, the client has no say on this matter.

**Comment by Pang654** [comment]
I am using Flask-Session which is a server-side session. The session id is sent in the cookie when the app is logged-in with HTTP request. So, the session id is stored in the client. To get the same session, I need to send the cookie back to the server, or the session id back to the server. I can able to access the same session if I have the session id.

As I know, there is no option for us to send the cookie from my socketio client, that's why i need to send back the session id through emit. Is there a way to update the server the session id with the emit message?

**Comment by miguelgrinberg** [comment]
A "sid" is a Socket.IO session id. It is part of the Socket.IO protocol. You are calling the Flask session id a sid, which is wrong. I thought you were talking about the real sid, my answer was for that.

You really mean the Flask session id, which is set in a cookie. That has to be sent in the connection request, so your client must send it. Unfortunately there is nothing that can be done from the server side, I suggest you modify your client to include the cookie header.

**Comment by Pang654** [comment]
Yup it works finally. Thanks.

**Comment by Bengab-coder** [comment]
Hi guys, Please I need help . I am building a forum website with flask and im using flask-socketio and flask-session for user management. I have also setup a javascript client code to connect to the server . Connection and message sending works perfectly fine but when I try to access the data in the session storage, im not seeing anything and if I also print it out when a user connects , it clearly showed that there was no data stored in the session storage or maybe it was cleared somehow, I set a key called "current_user_id" to be the user id of the currently logged in user. This works when I use normal http route but when I use websockets im not getting the expected results. 

### Here is the code for my server.

```
from flask import Flask,flash, redirect, render_template, request, url_for,session
from flask_socketio import SocketIO,send,SessionMixin
from flask_sqlalchemy import SQLAlchemy
from flask_session import Session
from datetime import datetime
from os import path
from helpers import is_allowed_file_extension

app = Flask(__name__)

app.config['SECRET_KEY'] = "youwillneverguessthiskey_flaskchatapp$%"
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///database.db'
app.config['SESSION_TYPE'] = 'filesystem'
Session(app)
socketio = SocketIO(app,cors_allowed_origins="*",manage_session=False)

db = SQLAlchemy(app)

class User(db.Model):
    id = db.Column(db.Integer,primary_key=True)
    username = db.Column(db.String(80),unique=True,nullable=False)
    fullname = db.Column(db.String(80),unique=True,nullable=False)
    password = db.Column(db.String(80),unique=False,nullable=False)
    path_to_profile_picture = db.Column(db.String(100),unique=True,nullable=False)

if not path.exists('database.db'):
    db.create_all()

@app.route("/")
def index():
    posts = Post.query.all()
    print("\n\nThis is what is in the sessions storage\n\n",session)
    current_id = session.get("current_user_id")
    return render_template("index.html",current_user_id = current_id,posts=posts)


@app.route("/login",methods=['GET','POST'])
def handle_login():

    if request.method == 'GET':
        return render_template("login_form.html")
    username = request.form.get('username')
    password = request.form.get('password')

    user = User.query.filter_by(username=username,password=password).first()
    
    if user:
        
        session['current_user_id'] = user.id
        flash(f"Successfully logged in as {session.get('current_user_id')}",category='success')
        return redirect(url_for('index'))
    else:
        flash(f"{username}  does not exists",category='error')
        return redirect(request.url)
    
@app.route("/logout")
def handle_logout():
    session.pop("current_user_id")
    flash("Logged out successfully",category='success')
    return redirect(url_for('index'))

@socketio.on('connect')
def handle_new_user_connection():
    print("\n\nA new user has connected, heres what's in the session storage",session,"\n\n")

@socketio.on('message')
def handle_message(received_data):
    try:
        new_post_text,msg_type = received_data
        if msg_type == "message":
            print("\n\nRan this line\n\n")
            print("Current User",session)
            
            ### The Problem starts here

            new_post = Post(body=new_post_text,owner_id=session["current_user_id"])
            db.session.add(new_post)
            db.session.commit()
            post_created_at = new_post.created_at.strftime("%d %B, %Y")

            data_to_send = [
                get_user_username_by_id(session["current_user_id"]),
                get_user_pic_path_by_id(session["current_useer"]),
                received_data,
                post_created_at,
                new_post.id
            ]

            send(data_to_send,broadcast=True)
            flash("Your post was published successfully",category='success')

    except ValueError:
        pass
```


### Also here is the client code, Im using jquery by the way


```
$(document).ready(function () {
    console.log("Page fully loaded");
    var socket = io.connect("http://127.0.0.1:5000/");
    socket.on('connect', () => {
        console.log("Connected");
        socket.send("A User has connected");
        
    });
    socket.on("message",function (msg) {
      console.log(msg)
        $("#posts-container-main-body").append(`
        <div class="post-item">
          <div class="post-header-container">
            <div class="post-item-info-header">
              <img class="profile-picture" src="${msg[1]}">
              <span>${msg[0]}</span>
            </div>
            <span class="date-label">${msg[3]}</span>
          </div>

          <h2 class="post-item-body">${msg[2][0]}</h2>
          // <button class="like-btn">
          //   <svg xmlns="http://www.w3.org/2000/svg" width="22" height="22" fill="currentColor" class="bi bi-hand-thumbs-up" viewBox="0 0 16 16">
          //     <path d="M8.864.046C7.908-.193 7.02.53 6.956 1.466c-.072 1.051-.23 2.016-.428 2.59-.125.36-.479 1.013-1.04 1.639-.557.623-1.282 1.178-2.131 1.41C2.685 7.288 2 7.87 2 8.72v4.001c0 .845.682 1.464 1.448 1.545 1.07.114 1.564.415 2.068.723l.048.03c.272.165.578.348.97.484.397.136.861.217 1.466.217h3.5c.937 0 1.599-.477 1.934-1.064a1.86 1.86 0 0 0 .254-.912c0-.152-.023-.312-.077-.464.201-.263.38-.578.488-.901.11-.33.172-.762.004-1.149.069-.13.12-.269.159-.403.077-.27.113-.568.113-.857 0-.288-.036-.585-.113-.856a2.144 2.144 0 0 0-.138-.362 1.9 1.9 0 0 0 .234-1.734c-.206-.592-.682-1.1-1.2-1.272-.847-.282-1.803-.276-2.516-.211a9.84 9.84 0 0 0-.443.05 9.365 9.365 0 0 0-.062-4.509A1.38 1.38 0 0 0 9.125.111L8.864.046zM11.5 14.721H8c-.51 0-.863-.069-1.14-.164-.281-.097-.506-.228-.776-.393l-.04-.024c-.555-.339-1.198-.731-2.49-.868-.333-.036-.554-.29-.554-.55V8.72c0-.254.226-.543.62-.65 1.095-.3 1.977-.996 2.614-1.708.635-.71 1.064-1.475 1.238-1.978.243-.7.407-1.768.482-2.85.025-.362.36-.594.667-.518l.262.066c.16.04.258.143.288.255a8.34 8.34 0 0 1-.145 4.725.5.5 0 0 0 .595.644l.003-.001.014-.003.058-.014a8.908 8.908 0 0 1 1.036-.157c.663-.06 1.457-.054 2.11.164.175.058.45.3.57.65.107.308.087.67-.266 1.022l-.353.353.353.354c.043.043.105.141.154.315.048.167.075.37.075.581 0 .212-.027.414-.075.582-.05.174-.111.272-.154.315l-.353.353.353.354c.047.047.109.177.005.488a2.224 2.224 0 0 1-.505.805l-.353.353.353.354c.006.005.041.05.041.17a.866.866 0 0 1-.121.416c-.165.288-.503.56-1.066.56z"/>
          //   </svg>
          //   <label class="like-count">80</label>
          // </button>

          // <button class="comment-btn">
          //   <svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" fill="currentColor" class="bi bi-chat-left" viewBox="0 0 16 16">
          //     <path d="M14 1a1 1 0 0 1 1 1v8a1 1 0 0 1-1 1H4.414A2 2 0 0 0 3 11.586l-2 2V2a1 1 0 0 1 1-1h12zM2 0a2 2 0 0 0-2 2v12.793a.5.5 0 0 0 .854.353l2.853-2.853A1 1 0 0 1 4.414 12H14a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2H2z"/>
          //   </svg>
          //   <label class="like-count">{{239}}</label>
          // </button>
            <a href="/post/${msg[4]}">
              <button>View Post</button>
            </a>
        </div>
        `);
    });
    $("#send-post-btn").on('click',function () {
        socket.send([$("#post-input-textarea").val(),'message']);
        $("#post-input-textarea").val('');
    });
});
```

**Comment by miguelgrinberg** [comment]
@Bengab-coder does [this example](https://github.com/miguelgrinberg/Flask-SocketIO/blob/main/example/sessions.py) work for you? I really can't debug your program with my eyes, but you have a working application that you can use to find the problem in your code.

**Comment by Bengab-coder** [comment]
> @Bengab-coder does [this example](https://github.com/miguelgrinberg/Flask-SocketIO/blob/main/example/sessions.py) work for you? I really can't debug your program with my eyes, but you have a working application that you can use to find the problem in your code.

Oh, I did not test your code example Sir , but I'm a bit confused, what about the client side to test the connection. Or can it still run well without the client connection?

**Comment by miguelgrinberg** [comment]
@Bengab-coder My example has back and front ends.

**Comment by Bengab-coder** [comment]
Okay please give me some time , I will reply back in some time let me check your example. Thanks for the help Sir.

**Comment by Bengab-coder** [comment]
> @Bengab-coder My example has back and front ends.

I think your example code works, I created a few routes to manually rest the session via normal http request and I also add a button and using javascript I set it up to send a message to the server  and so far so good, I think it works fine. Thanks for your help, ill comment if I have future issues but as for now my problem has been solved . Ill check my code to see where im making a mistake. Thank you Sir.
